subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy

```diff
@@ -510,4 +510,165 @@ dependencies { compile 'net.sf.ehcache:ehcache:2.10.2' }
+
+        where:
+        type << ['', 'static']
+    }
+
+    @Unroll("detects changes to class referenced through a #type array field")
+    def "detects changes to class referenced through an array field"() {
+        given:
+        java """class A {
+    $type B[] b;
+    void doSomething() {
+        Runnable r = b[0];
+        r.run();
+    }
+}"""
+        java '''abstract class B implements Runnable { }'''
+
+        outputs.snapshot { run "compileJava" }
+
+        when:
+        file("src/main/java/B.java").text = "class B { }"
+        fails "compileJava"
+
+        then:
+        errorOutput.contains 'Runnable r = b[0];'
+
+        where:
+        type << ['', 'static']
+    }
+
+    @Unroll("detects changes to class referenced through a #type multi-dimensional array field")
+    def "detects changes to class referenced through an multi-dimensional array field"() {
+        given:
+        java """class A {
+    $type B[][] b;
+    void doSomething() {
+        Runnable r = b[0][0];
+        r.run();
+    }
+}"""
+        java '''abstract class B implements Runnable { }'''
+
+        outputs.snapshot { run "compileJava" }
+
+        when:
+        file("src/main/java/B.java").text = "class B { }"
+        fails "compileJava"
+
+        then:
+        errorOutput.contains 'Runnable r = b[0][0];'
+
+        where:
+        type << ['', 'static']
+    }
+
+    def "detects changes to class referenced in method body"() {
+        given:
+        java '''class A {
+    void doSomething(Object b) {
+        Runnable r = (B) b;
+        r.run();
+    }
+}'''
+        java '''abstract class B implements Runnable { }'''
+
+        outputs.snapshot { run "compileJava" }
+
+        when:
+        file("src/main/java/B.java").text = "class B { }"
+        fails "compileJava"
+
+        then:
+        errorOutput.contains 'Runnable r = (B) b;'
+    }
+
+    def "detects changes to class referenced through return type"() {
+        given:
+        java '''class A {
+    B b() { return null; }
+    
+    void doSomething() {
+        Runnable r = b();
+        r.run();
+    }
+}'''
+        java '''abstract class B implements Runnable { }'''
+
+        outputs.snapshot { run "compileJava" }
+
+        when:
+        file("src/main/java/B.java").text = "class B { }"
+        fails "compileJava"
+
+        then:
+        errorOutput.contains 'Runnable r = b();'
+    }
+
+    def "detects changes to class referenced through method signature"() {
+        given:
+        java '''class A {
+    Runnable go(B b) {
+        Runnable r = b;
+        r.run();
+        return b;
+    }
+}'''
+        java '''abstract class B implements Runnable { }'''
+
+        outputs.snapshot { run "compileJava" }
+
+        when:
+        file("src/main/java/B.java").text = "class B { }"
+        fails "compileJava"
+
+        then:
+        errorOutput.contains 'Runnable r = b;'
+    }
+
+    def "detects changes to class referenced through type argument in field"() {
```

Comment:
I added the test cases. However, it is the same kind of problem as in gradle/performance#306. The test passes not because we analyze the type in the signature, but because it's actually used in the method body. If it wasn't, it would fail.